vnext/Shared/HermesRuntimeHolder.cpp

```diff
@@ -3,31 +3,108 @@
 
 #include "pch.h"
 
-#include <hermes/hermes.h>
+#include <memory>
 #include <mutex>
+
+#include <JSI/decorator.h>
+#include <cxxreact/MessageQueueThread.h>
+#include <cxxreact/SystraceSection.h>
+#include <hermes/hermes.h>
 #include "HermesRuntimeHolder.h"
 
+#if defined(HERMES_ENABLE_DEBUGGER)
+#include <hermes/inspector/chrome/Registration.h>
+#endif
+
 using namespace facebook;
 
 namespace facebook {
 namespace react {
 
+namespace {
+
+std::unique_ptr<facebook::hermes::HermesRuntime> makeHermesRuntimeSystraced(
+    const ::hermes::vm::RuntimeConfig &runtimeConfig) {
+  SystraceSection s("HermesExecutorFactory::makeHermesRuntimeSystraced");
+  return hermes::makeHermesRuntime(runtimeConfig);
+}
+
+#ifdef HERMES_ENABLE_DEBUGGER
+class HermesExecutorRuntimeAdapter : public facebook::hermes::inspector::RuntimeAdapter {
+ public:
+  HermesExecutorRuntimeAdapter(
+      std::shared_ptr<jsi::Runtime> runtime,
+      facebook::hermes::HermesRuntime &hermesRuntime,
+      std::shared_ptr<MessageQueueThread> thread)
+      : m_runtime(runtime), m_hermesRuntime(hermesRuntime), m_thread(std::move(thread)) {}
+
+  virtual ~HermesExecutorRuntimeAdapter() = default;
+
+  jsi::Runtime &getRuntime() override {
+    return *m_runtime;
+  }
+
+  facebook::hermes::debugger::Debugger &getDebugger() override {
+    return m_hermesRuntime.getDebugger();
+  }
+
+  void tickleJs() override {
+    // The queue will ensure that runtime_ is still valid when this
+    // gets invoked.
+    m_thread->runOnQueue([&runtime = m_runtime]() {
+      auto func = runtime->global().getPropertyAsFunction(*runtime, "__tickleJs");
+      func.call(*runtime);
+    });
+  }
+
+ private:
+  std::shared_ptr<jsi::Runtime> m_runtime;
+  facebook::hermes::HermesRuntime &m_hermesRuntime;
+
+  std::shared_ptr<MessageQueueThread> m_thread;
+};
+#endif
+
+} // namespace
+
 std::shared_ptr<jsi::Runtime> HermesRuntimeHolder::getRuntime() noexcept {
-  std::call_once(once_flag_, [this]() { initRuntime(); });
+  std::call_once(m_once_flag, [this]() { initRuntime(); });
 
-  if (!runtime_)
+  if (!m_runtime)
     std::terminate();
 
-  // ChakraJsiRuntime is not thread safe as of now.
-  if (own_thread_id_ != std::this_thread::get_id())
+  // Make sure that the runtime instance is not consumed from multiple threads.
+  if (m_own_thread_id != std::this_thread::get_id())
     std::terminate();
 
-  return runtime_;
+  return m_runtime;
 }
 
+HermesRuntimeHolder::HermesRuntimeHolder(
+    std::shared_ptr<facebook::react::DevSettings> devSettings,
+    std::shared_ptr<facebook::react::MessageQueueThread> jsQueue) noexcept
+    : m_devSettings(devSettings), m_jsQueue(jsQueue) {}
+
 void HermesRuntimeHolder::initRuntime() noexcept {
-  runtime_ = facebook::hermes::makeHermesRuntime();
-  own_thread_id_ = std::this_thread::get_id();
+  auto runtimeConfig = ::hermes::vm::RuntimeConfig();
```

Comment:
>::hermes::vm::RuntimeConfig(); [](http://example.com/codeflow?start=23&length=30)

It would be nice that RNW code does not access the Hermes private APIs directly. It can be part of a different PR as we dicussed yesterday to implement ABI-safe API.

Reply:
Sure.. Let's look at it when implementing ABI-safe APIs !